### DISCUSSION TITLE
Activity UI customization

### DISCUSSION THREAD

**Discussion by gabrySuerz** [elsa-workflows/elsa-core#4087] [Q&A] [answered]
Hi!

During the past few weeks I've tried few scenarios that I would like to implement with ELSA.

I've tested all requested scenarios but one, regarding a custom activity.
This are the steps that I did:
- I created a custom activity via C# (let's call it "MoveFile")
- I implemented the callback (es. moving a file) that will be called when the workflow is executed
- I configured ELSA to pickup also the activities from the executing assembly
- I opened the designer and added the "MoveFile" activity

Here is where I came to a blockage: I configured the activity with a custom property, but nothing was rendered.
The end goal would be to add fields such as dropdowns with a few custom choices.

It's present documentation on this scenario? Am I doing something wrong?

Thank you!

PS if needed I can attach the activity that I created.

**Comment by sfmskywalker** [comment, marked as answer]
Sure! There is *some* documentation about configuring your activity's properties [here](https://v2.elsaworkflows.io/docs/extensibility/custom-activities#properties), but admittedly, it's not very exhaustive.

Here's an example of an activity property that configures its UIto appear as a checklist, taken from the `HttpEndpoint` activity:

```csharp
/// <summary>
/// The HTTP methods that triggers this activity.
/// </summary>
[ActivityInput(
    UIHint = ActivityInputUIHints.CheckList,
    Hint = "The HTTP methods that trigger this activity.",
    Options = new[] { "GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "HEAD" },
    DefaultValueProvider = typeof(HttpEndpointDefaultMethodsProvider),
    DefaultSyntax = SyntaxNames.Json,
    SupportedSyntaxes = new[] { SyntaxNames.Json, SyntaxNames.JavaScript, SyntaxNames.Liquid })]

public HashSet<string> Methods { get; set; } = new() { "GET" };
```

Notice the use of the `UIHint` and `Options` attribute properties. Currently, these UI hints are supported:

```csharp
public static class ActivityInputUIHints
{
    public const string SingleLine = "single-line";
    public const string MultiLine = "multi-line";
    public const string Checkbox = "checkbox";
    public const string CheckList = "check-list";
    public const string RadioList = "radio-list";
    public const string Dropdown = "dropdown";
    public const string MultiText = "multi-text";
    public const string CodeEditor = "code-editor";
    public const string Dictionary = "dictionary";

    /// <summary>
    /// An editor that allows the user to write a blob of JSON.
    /// </summary>
    public const string Json = "json";
}
```

It's possible to implement your own UI hints, as documented [here](https://v2.elsaworkflows.io/docs/extensibility/designer-plugins#custom-field-types).

**Comment by gabrySuerz** [comment]
Thank you!

Before your answer, I admit that I didn't fully understand what the attribute did, now I do.
